### PR TITLE
Local api infra

### PR DESCRIPTION
**Pull request by liviuba**
https://app.clickup.com/t/8694jgg2e

Missing a few things, this is more to get feedback on the general direction

### CODE REVIEW

**Comment by ctr26** [comment]
First obvious thing that would be quite nice is the add GitHub actions testing for the helm chart (but not the infra deployment with helmsman). 

This is my GitHub action that I use for testing and listing helm charts. If you don't feel like it adding this in just yet that's fine I just thought you might want to see an example of what I do.

(I use chart testing, very useful: https://github.com/helm/chart-testing)

```
# See https://github.com/helm/chart-testing-action
name: Lint and Test Charts

on:
  push:
    branches:
      - main
  pull_request:
  workflow_dispatch:

jobs:
  lint-test:
    runs-on: ubuntu-latest
    # strategy:
      # matrix:
        # k8s: ["1.19.16", "1.21.14", "1.22.9", "1.23.12", "1.24.7", "1.25.2"]
    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with: 
          fetch-depth: 0

      - name: Install Helm
        uses: azure/setup-helm@v3
        # as of 2022/12 the set-output still not fixed in this action
        # https://github.com/Azure/setup-helm/issues/103
        with:
          version: v3.12.1

      - uses: actions/setup-python@v4
        with:
          python-version: '3.10'
          check-latest: true

      - name: Install chart-testing
        uses: helm/chart-testing-action@v2.6.0

      - name: Run chart-testing (list-changed)
        id: list-changed
        run: |
          changed=$(ct list-changed --target-branch ${{ github.event.repository.default_branch }})
          if [[ -n "$changed" ]]; then
            echo "changed=true" >> "$GITHUB_OUTPUT"
          fi

      - name: Run chart-testing (lint)
        if: steps.list-changed.outputs.changed == 'true'
        run: ct lint --config ct.yaml --target-branch ${{ github.event.repository.default_branch }} --all --debug 

      - name: Create kind cluster
        uses: helm/kind-action@v1.8.0
        with:
          config: kind-config.yaml  

      - name: Run chart-testing (install)
        if: steps.list-changed.outputs.changed == 'true'
        run: ct install --config ct.yaml --target-branch ${{ github.event.repository.default_branch }} --helm-extra-args "--timeout 600s"
         # See https://github.com/helm/chart-testing/blob/main/doc/ct_install.md
```

File: api/local/helmsman_local.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
+metadata:
```
Comment (ctr26) [inline api/local/helmsman_local.yml:1]: FYI for the future, In helmsman you can actually declare multi helmsman yamls which you can use to override settings. Very useful for maintaining local+dev+prod deployments

e.g. 
helmsman apply -f helmsman.yaml -f local.yaml

File: api/local/api/templates/NOTES.txt [hunk 2]
```diff
@@ -0,0 +1,22 @@
+1. Get the application URL by running these commands:
```
Comment (ctr26) [inline api/local/api/templates/NOTES.txt:1]: Very nice, I should do this with my helm charts!

File: api/local/api/templates/secrets.yaml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+apiVersion: v1
```
Comment (ctr26) [inline api/local/api/templates/secrets.yaml:1]: Extremely correct!

File: api/local/helmsman_local.yml [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    namespace: api
+    enabled: true
+    chart: "bitnami/mongodb"
+    version: "15.4.3"
+    priority: -2
+    set:
+      "auth.rootUser": $MONGO_ROOT_USER
+      "auth.rootPassword": $MONGO_ROOT_PASSWORD
+    valuesFiles:
+      - "mongo_values.yml"
```
Comment (ctr26) [inline api/local/helmsman_local.yml:22]: 👍

File: api/local/helmsman_local.yml [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  api:
+    namespace: api
+    enabled: true
+    chart: ./api
+    version: "0.1.0"
+    priority: -1
+    set:
```
Comment (ctr26) [inline api/local/helmsman_local.yml:29]: Great, really maintainable!

File: api/local/.env_template [hunk 6]
```diff
@@ -0,0 +1,6 @@
+MONGO_ROOT_USER=root
+MONGO_ROOT_PASSWORD=example
+# run: 'openssl rand -hex 32' to generate a secret key for testing
```
Comment (ctr26) [inline api/local/.env_template:3]: 👍

File: api/local/api/Chart.yaml [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+# This is the chart version. This version number should be incremented each time you make changes
+# to the chart and its templates, including the app version.
+# Versions are expected to follow Semantic Versioning (https://semver.org/)
+version: 0.1.0
+
+# This is the version number of the application being deployed. This version number should be
+# incremented each time you make changes to the application. Versions are not expected to
+# follow Semantic Versioning. They should reflect the version the application is using.
+# It is recommended to use it with quotes.
+appVersion: "1.16.0"
```
Comment (ctr26) [inline api/local/api/Chart.yaml:24]: This should be the version of the application (i.e the container image tag)

I usually then set my container image tag to default to the appVersion

File: api/local/README.md [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+# create RO classic token for minikube here https://github.com/settings/tokens
+kubectl create secret --namespace api docker-registry ghcr-creds \
+    --docker-server=https://ghcr.io \
+    --docker-username=GITHUB_USERNAME \
+    --docker-password=SEE_COMMENT \
+    --docker-email=GITHUB_EMAIL
+```
+
+* `cp .env_template .env`
+* helmsman --apply -f helmsman_local.yml -e .env
```
Comment (ctr26) [inline api/local/README.md:16]: 👍

File: api/local/api/templates/deployment.yaml [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+          image: "ghcr.io/bioimage-archive/integrator-api:{{ .Values.image.tag }}"
+          imagePullPolicy: {{ .Values.image.pullPolicy }}
+          resources:
+            {{- toYaml .Values.resources | nindent 12 }}
+          ports:
+            - name: http
+              containerPort: 8080
+              protocol: TCP
+          livenessProbe:
+            httpGet:
```
Comment (ctr26) [inline api/local/api/templates/deployment.yaml:29]: very useful!

**Comment by ctr26** [comment]
@liviuba this is very well done, you've made a lot of strong early design decisions here that will make the downstream really easy to maintain and manage! I can't find really anything of note that should be fixed or improved.

**Review by ctr26 — APPROVED** [review]

**Comment by ctr26** [comment]
@liviuba let me know if you intend on adding chart testing, otherwise I'm happy for it to be merged

**Comment by ctr26** [comment]
Could we change the path names of the helm chart to be a bit easier to work with please? Maybe 

api/helm-chart? for the helm chart

**Comment by liviuba** [comment]
> Could we change the path names of the helm chart to be a bit easier to work with please? Maybe
> 
> api/helm-chart? for the helm chart

Thanks! Moved things around so that the helm chart is rooted in api and then the local env stuff (wip) is separate, does this look better?

**Comment by ctr26** [comment]
Yep it does thanks

**Comment by ctr26** [comment]
Also (this is isn't urgent and it could just be done later) when I mentioned before helmsman I meant that more for the infrastructure deployment. What I would do here is use sub charts for mongo

https://github.com/bioimage-io/hypha-helm-charts/blob/main/charts/hypha/Chart.yaml
https://github.com/bioimage-io/hypha-helm-charts/blob/main/charts/hypha/values.yaml

And then in the values file you can set variables to the sub charts

**Comment by liviuba** [comment]
@ctr26 All done I think, can you have a look again please? Adding a new ticket to cleanup the test output

**Comment by ctr26** [comment]
I would maybe keep the docker tests in. The helm chart can only test against built and uploaded images (is this true?), so there could be a condition where a PR has happens that breaks the container but the tests don't catch it. 

If this isn't true then ignore me

**Review by ctr26 — APPROVED** [review]

**Pull request merged**